# Importing packages and df

In [153]:
#importando pacotes basicos para manipulação de df
import numpy as np
import pandas as pd
%matplotlib inline

In [154]:
wine = pd.read_csv('wines.csv')

Breve exploração dos dados

In [126]:
#checando os 5 primeiros dias de venda para todos os itens da Greenwich Village
wine.head()

date  ACCESORIES_1_001  ACCESORIES_1_002  ACCESORIES_1_003  \
0  2011-01-29               0.0               0.0               0.0   
1  2011-01-30               0.0               0.0               0.0   
2  2011-01-31               0.0               0.0               0.0   
3  2011-02-01               0.0               0.0               0.0   
4  2011-02-02               0.0               0.0               0.0   

   ACCESORIES_1_004  ACCESORIES_1_005  ACCESORIES_1_006  ACCESORIES_1_007  \
0          9.027410               0.0               0.0               0.0   
1         10.231064               0.0               0.0               0.0   
2          4.212791               0.0               0.0               0.0   
3         10.231064               0.0               0.0               0.0   
4          4.212791               0.0               0.0               0.0   

   ACCESORIES_1_008  ACCESORIES_1_009  ...  SUPERMARKET_3_818  \
0          4.492230          3.864971  ...                0.0   
1          3.288529          2.262400  ...                0.0   
2          3.533286          3.483516  ...                0.0   
3          2.907973          3.441201  ...                0.0   
4          3.777561          1.821771  ...                0.0   

   SUPERMARKET_3_819  SUPERMARKET_3_820  SUPERMARKET_3_821  SUPERMARKET_3_822  \
0          13.404426           4.252829                0.0          15.819579   
1           9.911063           3.069851                0.0          14.351716   
2           6.008622           3.072631                0.0          11.598456   
3           4.115047           2.833238                0.0          12.516661   
4           6.038693           1.180114                0.0           5.568878   

   SUPERMARKET_3_823  SUPERMARKET_3_824  SUPERMARKET_3_825  SUPERMARKET_3_826  \
0           5.014252           1.879593           2.964903                0.0   
1           3.010231           1.218895           7.409929                0.0   
2           3.019417           2.144103           1.976638                0.0   
3           4.693769           2.555720           6.416546                0.0   
4           1.328739           1.791521           3.952616                0.0   

   SUPERMARKET_3_827  
0                0.0  
1                0.0  
2                0.0  
3                0.0  
4                0.0  

[5 rows x 3050 columns]

In [155]:
#transformando a data para o formato date time e usando como indice para nosso df
wine.set_index('date', inplace=True)
wine.index = pd.to_datetime(wine.index)

In [156]:
wine

ACCESORIES_1_001  ACCESORIES_1_002  ACCESORIES_1_003  \
date                                                               
2011-01-29          0.000000          0.000000          0.000000   
2011-01-30          0.000000          0.000000          0.000000   
2011-01-31          0.000000          0.000000          0.000000   
2011-02-01          0.000000          0.000000          0.000000   
2011-02-02          0.000000          0.000000          0.000000   
...                      ...               ...               ...   
2016-04-20          8.803256          2.112040          0.395010   
2016-04-21          6.609972          1.056020          0.793264   
2016-04-22          3.318104          1.056020          2.370060   
2016-04-23          9.914108          1.580300          1.185030   
2016-04-24         12.125646          3.167588          1.573612   

            ACCESORIES_1_004  ACCESORIES_1_005  ACCESORIES_1_006  \
date                                                               
2011-01-29          9.027410          0.000000          0.000000   
2011-01-30         10.231064          0.000000          0.000000   
2011-01-31          4.212791          0.000000          0.000000   
2011-02-01         10.231064          0.000000          0.000000   
2011-02-02          4.212791          0.000000          0.000000   
...                      ...               ...               ...   
2016-04-20          3.607895          2.283050          1.185451   
2016-04-21          4.815313          3.418527          1.053529   
2016-04-22          9.022925          2.345321          0.526873   
2016-04-23         20.954232          7.224695          0.396053   
2016-04-24          8.973170          5.391122          2.114629   

            ACCESORIES_1_007  ACCESORIES_1_008  ACCESORIES_1_009  \
date                                                               
2011-01-29          0.000000          4.492230          3.864971   
2011-01-30          0.000000          3.288529          2.262400   
2011-01-31          0.000000          3.533286          3.483516   
2011-02-01          0.000000          2.907973          3.441201   
2011-02-02          0.000000          3.777561          1.821771   
...                      ...               ...               ...   
2016-04-20          5.244834          2.339566          0.469944   
2016-04-21          5.235898          3.668144          0.681871   
2016-04-22          0.000000          2.024934          0.216280   
2016-04-23          6.294386          2.274022          1.546797   
2016-04-24          3.145491          2.213650          2.522447   

            ACCESORIES_1_010  ...  SUPERMARKET_3_818  SUPERMARKET_3_819  \
date                          ...                                         
2011-01-29          1.976996  ...             0.0000          13.404426   
2011-01-30          2.373397  ...             0.0000           9.911063   
2011-01-31          1.978294  ...             0.0000           6.008622   
2011-02-01          1.582357  ...             0.0000           4.115047   
2011-02-02          1.186235  ...             0.0000           6.038693   
...                      ...  ...                ...                ...   
2016-04-20          0.000000  ...            12.8736           4.064614   
2016-04-21          0.791132  ...             7.8672           2.882281   
2016-04-22          3.166475  ...            12.5160           1.699075   
2016-04-23          3.559631  ...            13.5888           3.826988   
2016-04-24          4.352617  ...            10.3704           5.513501   

            SUPERMARKET_3_820  SUPERMARKET_3_821  SUPERMARKET_3_822  \
date                                                                  
2011-01-29           4.252829           0.000000          15.819579   
2011-01-30           3.069851           0.000000          14.351716   
2011-01-31           3.072631           0.000000          11.598456   
2011-02-01           2.833

In [157]:
# divisao entre treino e validacao
from sklearn.model_selection import train_test_split

train, test = train_test_split(wine, shuffle=False, test_size=12)

In [158]:
train.tail()

ACCESORIES_1_001  ACCESORIES_1_002  ACCESORIES_1_003  \
date                                                               
2016-04-08          0.000000          1.056020          0.787909   
2016-04-09          5.516100          2.111568          0.395010   
2016-04-10          5.503000          1.056020          1.573612   
2016-04-11          4.408871          0.528010          0.793264   
2016-04-12          1.107229          1.056020          1.186163   

            ACCESORIES_1_004  ACCESORIES_1_005  ACCESORIES_1_006  \
date                                                               
2016-04-08         10.806509          3.842520          2.134319   
2016-04-09         16.743379          3.509902          0.263516   
2016-04-10         12.582282          3.799922          1.454106   
2016-04-11          8.399200          5.003561          0.930751   
2016-04-12          9.028270          1.517822          1.203422   

            ACCESORIES_1_007  ACCESORIES_1_008  ACCESORIES_1_009  \
date                                                               
2016-04-08          3.145491          1.518825          0.234534   
2016-04-09          7.337268          2.909662          2.299399   
2016-04-10          4.196254          2.090925          1.392584   
2016-04-11          2.097338          2.781911          3.243563   
2016-04-12          6.292269          1.769845          3.942496   

            ACCESORIES_1_010  ...  SUPERMARKET_3_818  SUPERMARKET_3_819  \
date                          ...                                         
2016-04-08          2.374416  ...            10.3704           3.137653   
2016-04-09          2.769148  ...             7.8672           8.008374   
2016-04-10          1.582264  ...             5.7216           4.066160   
2016-04-11          3.165085  ...             2.1456           2.910155   
2016-04-12          2.768963  ...             2.1456           5.537603   

            SUPERMARKET_3_820  SUPERMARKET_3_821  SUPERMARKET_3_822  \
date                                                                  
2016-04-08           6.148231             1.7928          13.978227   
2016-04-09           7.339313             6.5736          18.600268   
2016-04-10           5.443911             5.9760          16.717196   
2016-04-11           3.311987             0.5976           9.346185   
2016-04-12           5.919880             2.9880          13.431629   

            SUPERMARKET_3_823  SUPERMARKET_3_824  SUPERMARKET_3_825  \
date                                                                  
2016-04-08           0.000000           1.549162           5.415202   
2016-04-09           0.000000           1.196994           7.390419   
2016-04-10           0.000000           0.300949           5.390767   
2016-04-11           2.341978           1.182217           5.434611   
2016-04-12           4.366217           2.144103           6.403692   

            SUPERMARKET_3_826  SUPERMARKET_3_827  
date                                              
2016-04-08           2.150400           2.879478  
2016-04-09           2.303605           2.640000  
2016-04-10           2.457007           5.039478  
2016-04-11           2.764405           2.878957  
2016-04-12           1.228800           3.120000  

[5 rows x 3049 columns]

# Modelo de previsão global (usando as 3049 séries temporais respectiva de cada produto vendido na Greenwich Village)

In [159]:
from src.tde import UnivariateTDE

In [160]:
# desta vez fazemos a divisão no tempo
train_end = pd.Timestamp('2015-06-01')
test_start = pd.Timestamp('2016-02-01')

In [161]:
training, testing, mean_value = {}, {}, {}
for col in wine.columns:
    # últimos 6 meses para validação
    train = wine.loc[:train_end, col]
    test = wine.loc[test_start:, col]

    # normalização com dados de TREINO!
    mean_value[col] = train.mean()
    train /= mean_value[col]
    test /= mean_value[col]

    # serie temporal como matriz
    training[col] = UnivariateTDE(train, horizon=1, k=4)
    testing[col] = UnivariateTDE(test, horizon=1, k=4)

In [162]:
training['ACCESORIES_1_001'].head()

t-3  t-2  t-1  t-0  t+1
0  0.0  0.0  0.0  0.0  0.0
1  0.0  0.0  0.0  0.0  0.0
2  0.0  0.0  0.0  0.0  0.0
3  0.0  0.0  0.0  0.0  0.0
4  0.0  0.0  0.0  0.0  0.0

In [163]:
# concatenando as matrizes
train_df = pd.concat(training, axis=0)
train_df.head()

t-3  t-2  t-1  t-0  t+1
ACCESORIES_1_001 0  0.0  0.0  0.0  0.0  0.0
                 1  0.0  0.0  0.0  0.0  0.0
                 2  0.0  0.0  0.0  0.0  0.0
                 3  0.0  0.0  0.0  0.0  0.0
                 4  0.0  0.0  0.0  0.0  0.0

In [164]:
train_df.tail()

t-3       t-2       t-1       t-0       t+1
SUPERMARKET_3_827 1576  3.339833  3.009775  5.013384  2.675356  4.678965
                  1577  3.009775  5.013384  2.675356  4.678965  3.006867
                  1578  5.013384  2.675356  4.678965  3.006867  6.353970
                  1579  2.675356  4.678965  3.006867  6.353970  2.340936
                  1580  4.678965  3.006867  6.353970  2.340936  8.691998

In [165]:
testing['ACCESORIES_1_001'].head()

t-3       t-2       t-1       t-0       t+1
0  1.931765  3.864294  2.571141  4.498302  8.350334
1  3.864294  2.571141  4.498302  8.350334  3.218874
2  2.571141  4.498302  8.350334  3.218874  2.569840
3  4.498302  8.350334  3.218874  2.569840  2.570540
4  8.350334  3.218874  2.569840  2.570540  2.577055

# Treinando modelo:

In [166]:
from sklearn.linear_model import Ridge
# treinando o modelo global
global_model = Ridge()
global_model.fit(X_train, y_train)

Ridge()

# Previsão:

In [169]:
# previsão
X_test = testing['ACCESORIES_1_001'].drop('t+1', axis=1)
local_forecasts = local_model.predict(X_test)

local_forecasts *= mean_value['ACCESORIES_1_001']

In [170]:
from sklearn.metrics import mean_absolute_error

In [171]:
mean_absolute_error(testing['ACCESORIES_1_001']['t+1'], global_forecasts)

2.4830728336756716

In [172]:
mean_absolute_error(testing['ACCESORIES_1_001']['t+1'], local_forecasts)

2.57105932176796

In [173]:
# prever o próximo valor com base nos 6 anteriores
series_tde = UnivariateTDE(wine['ACCESORIES_1_001'].diff(), horizon=1, k=6)
series_tde.head()

t-5  t-4  t-3  t-2  t-1  t-0  t+1
0  0.0  0.0  0.0  0.0  0.0  0.0  0.0
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0
2  0.0  0.0  0.0  0.0  0.0  0.0  0.0
3  0.0  0.0  0.0  0.0  0.0  0.0  0.0
4  0.0  0.0  0.0  0.0  0.0  0.0  0.0

In [174]:
# separar variáveis explicativas da variável objetivo
X = series_tde.drop('t+1', axis=1)
y = series_tde['t+1']

In [175]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, shuffle=False)

In [176]:
X_train.head()

t-5  t-4  t-3  t-2  t-1  t-0
0  0.0  0.0  0.0  0.0  0.0  0.0
1  0.0  0.0  0.0  0.0  0.0  0.0
2  0.0  0.0  0.0  0.0  0.0  0.0
3  0.0  0.0  0.0  0.0  0.0  0.0
4  0.0  0.0  0.0  0.0  0.0  0.0

In [177]:
y_train.head()

0    0.0
1    0.0
2    0.0
3    0.0
4    0.0
Name: t+1, dtype: float64

ESTIMAÇÂO DE DESEMPENHo


In [178]:
# erro estimado em validação cruzada
print(f'Average error is: {np.mean(cv_err)}')

Average error is: 0.5469058411212515


In [179]:
# erro que obtemos no teste
model = Ridge()
model.fit(X_train, y_train)
pred_test = model.predict(X_test)
err_test = mean_absolute_error(y_test, pred_test)
print(f'Test error is: {err_test}')

Test error is: 1.721970007449671
